File: js/problem15.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */


var res = null;


function twoSum(nums, start) {
    let i = start + 1, j = nums.length - 1;
    while (i < j) {
        
        let temp = nums[start] + nums[i] + nums[j];
        if (temp === 0) {
            res.push([nums[start], nums[i], nums[j]]);
            i++;
            while ((i < j) && (nums[i] === nums[i - 1])) {
                i++;
            }
            j--;
            while ((i < j) && (nums[j] === nums[j + 1])) {
                j--;
            }
        } else if (temp < 0) {
            i++;
        } else {
            j--;
        }
    }
}

var threeSum = function(nums) {
    
    if (nums.length < 3) {
        return [];
    }
    
    res = [];
    
    nums = nums.sort(function(a, b) {
        return (+a) - (+b);
    });
    
    for (let i = 0; i < nums.length - 2; i++) {
        if ((i > 0) && (nums[i] === nums[i - 1])) {
            continue;
        }
        
        twoSum(nums, i);
    }
    
    return res;
};

File: leetcodeJava/src/main/java/Problem15.java
// 15. 3Sum

class Solution {
    
    private void twoSum(int[] nums, int current, List<List<Integer>> res) {
        int i = current + 1, j = nums.length - 1;
        while (i < j) {
            int tempSum = nums[current] + nums[i] + nums[j];
            if (tempSum == 0) {
                List<Integer> tempList = new ArrayList<Integer>();
                tempList.add(nums[current]);
                tempList.add(nums[i]);
                tempList.add(nums[j]);
                res.add(tempList);
                i++;
                while ((i < j) && (nums[i] == nums[i - 1])) {
                    i++;
                }
                j--;
                while ((i < j) && (nums[j] == nums[j + 1])) {
                    j--;
                }
            } else if (tempSum < 0) {
                i++;
            } else {
                j--;
            }
        }
    }
    
    public List<List<Integer>> threeSum(int[] nums) {
        if (nums.length < 3) {
            return new ArrayList<List<Integer>>();
        } else if (nums.length == 3) {
            if(nums[0] + nums[1] + nums[2] == 0) {
                List<Integer> temp = new ArrayList<Integer>();
                for (int n : nums) {
                    temp.add(n);
                }
                List<List<Integer>> res = new ArrayList<List<Integer>>();
                res.add(temp);
                return res;
            }
        }
        
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        for (int i = 0; i < nums.length - 2; i++) {
            if ((i > 0) && (nums[i] == nums[i - 1])) {
                continue;
            }
            this.twoSum(nums, i, res);
        }
        
        return res;
    }
}